documentation/jetty-documentation/src/main/java/org/eclipse/jetty/docs/programming/client/http/HTTPClientDocs.java

```diff
@@ -919,57 +929,134 @@ public void dynamicOneProtocol()
         // end::dynamicOneProtocol[]
     }
 
-    public void dynamicH1H2() throws Exception
+    public void dynamicH1H2H3() throws Exception
     {
-        // tag::dynamicH1H2[]
+        // tag::dynamicH1H2H3[]
+        SslContextFactory.Client sslContextFactory = new SslContextFactory.Client();
+
         ClientConnector connector = new ClientConnector();
+        connector.setSslContextFactory(sslContextFactory);
 
         ClientConnectionFactory.Info http1 = HttpClientConnectionFactory.HTTP11;
 
         HTTP2Client http2Client = new HTTP2Client(connector);
         ClientConnectionFactoryOverHTTP2.HTTP2 http2 = new ClientConnectionFactoryOverHTTP2.HTTP2(http2Client);
 
-        HttpClientTransportDynamic transport = new HttpClientTransportDynamic(connector, http1, http2);
+        ClientQuicConfiguration quicConfiguration = new ClientQuicConfiguration(sslContextFactory, null);
+        HTTP3Client http3Client = new HTTP3Client(quicConfiguration, connector);
+        ClientConnectionFactoryOverHTTP3.HTTP3 http3 = new ClientConnectionFactoryOverHTTP3.HTTP3(http3Client);
+
+        // The order of the protocols indicates the client's preference.
```

Comment:
Can you please clarify whether the order is from the most preferred to least preferred, or the other way around? In the following line, does the client prefer HTTP/1 first, HTTP/2 second and HTTP/3 third?

Is it technically possible to flip that around and have it try HTTP/3, followed by HTTP/2, followed by HTTP/1? Or is this not possible because the connection type needs to be upgraded through the earlier versions?

Reply:
@cowwoc clarified the preference.

It is possible to reverse the order. 

However, since HTTP/3 is not compatible with HTTP/2 and HTTP/1, there is no way to "downgrade" the attempt (and switch from UDP to TCP), so if the server does not support HTTP/3 the request will fail.
Applications may handle the failure and attempt to specify explicitly the HTTP version via `request.version(...)`.
This is not done automatically by `HttpClient`.

The reverse order may be useful if you want your default to be HTTP/3, but you want to explicitly specify HTTP/2 or HTTP/1 for those servers you know don't support HTTP/3.

Reply:
@sbordet if the order of preference is HTTP/1, HTTP/2, HTTP/3 does that mean that the client will settle on HTTP/1 if the server supports all three protocols? Or will it try to upgrade the connection to HTTP/3?

I suspect that the client will settle on HTTP/1. If this is the case, how could one configure the client to try HTTP/3 and if that fails try HTTP/1 followed by an upgrade to HTTP/2? And if the upgrade to HTTP/2 fails then stay on HTTP/1.

I may be wrong, but this sounds like the happy path to me.

Reply:
HTTP/2 does not usually work with upgrades, it's negotiated via TLS + ALPN.

A client configured with `http/1.1, h2` will send that to the server, and the server will likely pick `h2` because it's the server that decides.

Trying HTTP/3 with a server that does not support it will result in a failure after the connect timeout (differently from TCP, where the failure is almost immediate).
After that timeout, the application may try explicitly HTTP/2.

That's why it's probably most common to do the opposite: if you know the server supports HTTP/3, you explicitly try that; otherwise you are not explicit and therefore HTTP/1.1+HTTP/2 will be negotiated.

Likely, as of now, the best configuration for the client is therefore: HTTP/2, HTTP/1.1, HTTP/3.

Reply:
We can't use the DNS extension as it is not available from Java.

About `Alt-Svc`, you have to think of `HttpClient` as `curl`: I doubt that if you make a request with curl saying "use HTTP/2", then curl will store the fact that there was an `Alt-Svc` header and the next request will be done via HTTP/3.

Applications can do that logic, because I guess whatever we do it will break someone.

For example, we do not want to do "preflight" requests to figure out `Alt-Svc` under the covers, as people will expect `HttpClient` to do exactly the request that they are telling it to do, not more.
Some other people will be happy if we do, and we don't want to have an option for all possible combinations of "how to discover HTTP/3".
Some other people won't care about HTTP/3, and really want to use HTTP/1.1 or HTTP/2.
A user can configure the dynamic transport with preference for HTTP/1.1, so it really wants to use HTTP/1.1, even if the server supports HTTP/3.
Just because HTTP/3 is the latest spec, does not mean people want to use it: it is not implicitly better than HTTP/1.1 in this context (it may be for browsers).

`HttpClient` is lower level than browsers, whose job is to show pages no matter how they are retrieved.

@cowwoc I am not sure whether you are asking for a feature here, or just discussing ideas.
I'm happy to discuss either.

Reply:
> Meaning, the first connection would try HTTP/2, HTTP/1.1, HTTP/3 but if HTTP/3 support is found then subsequent connections to the same host would try HTTP/3 first.

That would be possible, I guess. Need to parse the `Alt-Svc` header or the [`ALTSVC` frame](https://datatracker.ietf.org/doc/html/rfc7838#section-7.2), and then the `HttpClientTransportDynamic` could match scheme+authority on the destinations when the HTTP version is not explicitly set in the request, and pull out the preferred `Destination` to use.

As I said, that implies an order relationship based on the protocol (`HTTP/3.compareTo(HTTP/2) > 0`) that may be true for browsers, but not for `HttpClient`.

@cowwoc if you really want this, please file a different issue, because it is out of scope for this PR. Thanks!

Reply:
@sbordet I guess the only thing I'm wondering is whether:

1. I can start out with an HTTP/2, HTTP/1.1, HTTP/3 preference.
2. Upon making the first call to a hostname, take note whether the server response contains a `Alt-Svc: h3="<port number>` header.
3. Configure Jetty to use a different order of preference for that host in future calls.

If Jetty doesn't provide a way to set preferences on a per-host basis, one thing I could do is create one `HttpClient` per preference order (one normal, one that prefers HTTP/3) and use application-level logic to issue requests with one or the other depending on whether the host is known to support HTTP/3.

Reply:
Bullet 3 should be: Explicitly set the HTTP version on the request via `request.version(HttpVersion.HTTP_3)`.

That's it.

The option of configuring different `HttpClient` instances is not one I would suggest.

Reply:
I got HTTPS/3 working on the server side (works from a browser) but I can't get past the dreaded PKIX path building failed exception. Do you any documentation or test you could point me to showing how to get this working on the client side? I've got all my server-side certificates and private key but I'm unable to make the client-side accept the certificate root...

Reply:
@sbordet It looks like this PR is fine. See https://github.com/jetty/jetty.project/issues/11417 for a follow-up discussion.